File: sdk/python/symbolchain/facade/SymbolFacade.py
import hashlib
from datetime import datetime, timezone

from .. import sc
from ..CryptoTypes import Hash256, PublicKey, Signature
from ..Network import NetworkLocator
from ..symbol.KeyPair import KeyPair, Verifier
from ..symbol.Merkle import MerkleHashBuilder
from ..symbol.MessageEncoder import MessageEncoder
from ..symbol.Network import Address, Network
from ..symbol.SharedKey import SharedKey
from ..symbol.TransactionFactory import TransactionFactory

TRANSACTION_HEADER_SIZE = sum(field[1] for field in [
	('size', 4),
	('reserved1', 4),
	('signature', Signature.SIZE),
	('signer', PublicKey.SIZE),
	('reserved2', 4)
])

PRE_V3_AGGREGATE_HASHED_SIZE = sum(field[1] for field in [
	('version_network_type', 4),
	('max_fee', 8),
	('deadline', 8),
	('transactions_hash', Hash256.SIZE)
])

AGGREGATE_HASHED_SIZE = PRE_V3_AGGREGATE_HASHED_SIZE + sum(field[1] for field in [
	('payload_size', 4)
])


# region SymbolPublicAccount / SymbolAccount

class SymbolPublicAccount:
	"""Symbol public account."""

	def __init__(self, facade, public_key):
		"""Creates a Symbol public account."""
		self._facade = facade
		self.public_key = public_key
		self.address = self._facade.network.public_key_to_address(self.public_key)


class SymbolAccount(SymbolPublicAccount):
	"""Symbol account."""

	def __init__(self, facade, key_pair):
		"""Creates a Symbol account."""
		super().__init__(facade, key_pair.public_key)
		self.key_pair = key_pair

	def message_encoder(self):
		"""Creates a message encoder that can be used for encrypting and encoding messages between two parties."""
		return MessageEncoder(self.key_pair)

	def sign_transaction(self, transaction):
		"""Signs a Symbol transaction."""
		return self._facade.sign_transaction(self.key_pair, transaction)

	def cosign_transaction(self, transaction, detached=False):
		"""Cosigns a Symbol transaction."""
		return self._facade.cosign_transaction(self.key_pair, transaction, detached)

	def cosign_transaction_hash(self, transaction_hash, detached=False):
		"""Cosigns a Symbol transaction hash."""
		return self._facade.cosign_transaction_hash(self.key_pair, transaction_hash, detached)

# endregion


class SymbolFacade:
	"""Facade used to interact with Symbol blockchain."""

	BIP32_CURVE_NAME = 'ed25519'

	Address = Address  # pylint: disable=duplicate-code
	KeyPair = KeyPair  # pylint: disable=duplicate-code
	Verifier = Verifier
	SharedKey = SharedKey

	def __init__(self, network, account_descriptor_repository=None):
		"""Creates a Symbol facade."""
		self.network = NetworkLocator.find_by_name(Network.NETWORKS, network) if isinstance(network, str) else network
		self.account_descriptor_repository = account_descriptor_repository
		self.transaction_factory = self._create_symbol_transaction_factory()

	def _create_symbol_transaction_factory(self):
		type_parsing_rules = None
		if self.account_descriptor_repository:
			type_to_property_mapping = self._create_symbol_type_to_property_mapping()
			type_parsing_rules = self.account_descriptor_repository.to_type_parsing_rules_map(type_to_property_mapping)

		return TransactionFactory(self.network, type_parsing_rules)

	# NOTE: currently `TypeParserBuilder.create_sdk_wrapper`` assumes SDK types are used as keys (although could be `sc` types as well)
	# resulting sdk type will be converted further via `symbol_type_converter`
	@staticmethod
	def _create_symbol_type_to_property_mapping():
		return {
			Address: 'address',
			PublicKey: 'public_key',
		}

	def now(self):
		"""Creates a network timestamp representing the current time."""
		return self.network.from_datetime(datetime.now(timezone.utc))

	def create_public_account(self, public_key):
		"""Creates a Symbol public account from a public key."""
		return SymbolPublicAccount(self, public_key)

	def create_account(self, private_key):
		"""Creates a Symbol account from a private key."""
		return SymbolAccount(self, KeyPair(private_key))

	def hash_transaction(self, transaction):
		"""Hashes a Symbol transaction."""
		hasher = hashlib.sha3_256()
		hasher.update(transaction.signature.bytes)
		hasher.update(transaction.signer_public_key.bytes)
		hasher.update(self.network.generation_hash_seed.bytes)
		hasher.update(self._transaction_data_buffer(transaction.serialize()))
		return Hash256(hasher.digest())

	def extract_signing_payload(self, transaction):
		"""Gets the payload to sign given a Symbol transaction."""
		sign_buffer = self.network.generation_hash_seed.bytes
		sign_buffer += self._transaction_data_buffer(transaction.serialize())
		return sign_buffer

	def sign_transaction(self, key_pair, transaction):
		"""Signs a Symbol transaction."""
		return key_pair.sign(self.extract_signing_payload(transaction))

	def verify_transaction(self, transaction, signature):
		"""Verifies a Symbol transaction."""
		return Verifier(transaction.signer_public_key).verify(self.extract_signing_payload(transaction), signature)

	@staticmethod
	def cosign_transaction_hash(key_pair, transaction_hash, detached=False):
		"""Cosigns a Symbol transaction hash."""
		cosignature = sc.DetachedCosignature() if detached else sc.Cosignature()
		if detached:
			cosignature.parent_hash = sc.Hash256(transaction_hash.bytes)

		cosignature.version = 0
		cosignature.signer_public_key = sc.PublicKey(key_pair.public_key.bytes)
		cosignature.signature = sc.Signature(key_pair.sign(transaction_hash.bytes).bytes)
		return cosignature

	def cosign_transaction(self, key_pair, transaction, detached=False):
		"""Cosigns a Symbol transaction."""
		transaction_hash = self.hash_transaction(transaction)
		return self.cosign_transaction_hash(key_pair, transaction_hash, detached)

	@staticmethod
	def hash_embedded_transactions(embedded_transactions):
		"""Hashes embedded transactions of an aggregate."""
		hash_builder = MerkleHashBuilder()
		for embedded_transaction in embedded_transactions:
			hash_builder.update(Hash256(hashlib.sha3_256(embedded_transaction.serialize()).digest()))

		return hash_builder.final()

	def bip32_path(self, account_id):
		"""Creates a network compatible BIP32 path for the specified account."""
		return [44, 4343 if 'mainnet' == self.network.name else 1, account_id, 0, 0]

	@staticmethod
	def bip32_node_to_key_pair(bip32_node):
		"""Derives a Symbol KeyPair from a BIP32 node."""
		return KeyPair(bip32_node.private_key)

	@staticmethod
	def _is_aggregate_transaction(transaction_buffer):
		transaction_type_offset = TRANSACTION_HEADER_SIZE + 2  # skip version and network byte
		transaction_type = (transaction_buffer[transaction_type_offset + 1] << 8) + transaction_buffer[transaction_type_offset]
		aggregate_types = [sc.TransactionType.AGGREGATE_BONDED.value, sc.TransactionType.AGGREGATE_COMPLETE.value]
		return transaction_type in aggregate_types

	@staticmethod
	def _transaction_data_buffer(transaction_buffer):
		data_buffer_start = TRANSACTION_HEADER_SIZE
		data_buffer_end = len(transaction_buffer)
		if SymbolFacade._is_aggregate_transaction(transaction_buffer):
			version = transaction_buffer[TRANSACTION_HEADER_SIZE]
			data_buffer_end = TRANSACTION_HEADER_SIZE + (AGGREGATE_HASHED_SIZE if version >= 3 else PRE_V3_AGGREGATE_HASHED_SIZE)

		return transaction_buffer[data_buffer_start:data_buffer_end]

File: sdk/javascript/src/facade/SymbolFacade.js
/* eslint-disable no-unused-vars */
import { Bip32Node } from '../Bip32.js';
/* eslint-enable no-unused-vars */
import {
	Hash256,
	PrivateKey,
	PublicKey,
	/* eslint-disable no-unused-vars */
	SharedKey256,
	/* eslint-enable no-unused-vars */
	Signature
} from '../CryptoTypes.js';
import { NetworkLocator } from '../Network.js';
import { KeyPair, Verifier } from '../symbol/KeyPair.js';
import MessageEncoder from '../symbol/MessageEncoder.js';
import {
	Address,
	Network,
	/* eslint-disable no-unused-vars */
	NetworkTimestamp
	/* eslint-enable no-unused-vars */
} from '../symbol/Network.js';
import { deriveSharedKey } from '../symbol/SharedKey.js';
import TransactionFactory from '../symbol/TransactionFactory.js';
import { MerkleHashBuilder } from '../symbol/merkle.js';
import * as sc from '../symbol/models.js';
import { sha3_256 } from '@noble/hashes/sha3.js';

const TRANSACTION_HEADER_SIZE = [
	4, // size
	4, // reserved1
	Signature.SIZE, // signature
	PublicKey.SIZE, // signer
	4 // reserved2
].reduce((x, y) => x + y);

const PRE_V3_AGGREGATE_HASHED_SIZE = [
	4, // version, network, type
	8, // maxFee
	8, // deadline
	Hash256.SIZE // transactionsHash
].reduce((x, y) => x + y);

const AGGREGATE_HASHED_SIZE = PRE_V3_AGGREGATE_HASHED_SIZE + [
	4 // payloadSize
].reduce((x, y) => x + y);

const isAggregateTransaction = transactionBuffer => {
	const transactionTypeOffset = TRANSACTION_HEADER_SIZE + 2; // skip version and network byte
	const transactionType = (transactionBuffer[transactionTypeOffset + 1] << 8) + transactionBuffer[transactionTypeOffset];
	const aggregateTypes = [sc.TransactionType.AGGREGATE_BONDED.value, sc.TransactionType.AGGREGATE_COMPLETE.value];
	return aggregateTypes.some(aggregateType => aggregateType === transactionType);
};

const transactionDataBuffer = transactionBuffer => {
	const dataBufferStart = TRANSACTION_HEADER_SIZE;
	let dataBufferEnd = transactionBuffer.length;
	if (isAggregateTransaction(transactionBuffer)) {
		const version = transactionBuffer[TRANSACTION_HEADER_SIZE];
		dataBufferEnd = TRANSACTION_HEADER_SIZE + (3 <= version ? AGGREGATE_HASHED_SIZE : PRE_V3_AGGREGATE_HASHED_SIZE);
	}

	return transactionBuffer.subarray(dataBufferStart, dataBufferEnd);
};

// region SymbolPublicAccount / SymbolAccount

/**
 * Symbol public account.
 */
export class SymbolPublicAccount {
	/**
	 * Creates a Symbol public account.
	 * @param {SymbolFacade} facade Symbol facade.
	 * @param {PublicKey} publicKey Account public key.
	 */
	constructor(facade, publicKey) {
		/**
		 * @protected
		 */
		this._facade = facade;

		/**
		 * Account public key.
		 * @type {PublicKey}
		 */
		this.publicKey = publicKey;

		/**
		 * Account address.
		 * @type {Address}
		 */
		this.address = this._facade.network.publicKeyToAddress(this.publicKey);
	}
}

/**
 * Symbol account.
 */
export class SymbolAccount extends SymbolPublicAccount {
	/**
	 * Creates a Symbol account.
	 * @param {SymbolFacade} facade Symbol facade.
	 * @param {KeyPair} keyPair Account key pair.
	 */
	constructor(facade, keyPair) {
		super(facade, keyPair.publicKey);

		/**
		 * Account key pair.
		 * @type {KeyPair}
		 */
		this.keyPair = keyPair;
	}

	/**
	 * Creates a message encoder that can be used for encrypting and encoding messages between two parties.
	 * @returns {MessageEncoder} Message encoder using this account as one party.
	 */
	messageEncoder() {
		return new MessageEncoder(this.keyPair);
	}

	/**
	 * Signs a Symbol transaction.
	 * @param {sc.Transaction} transaction Transaction object.
	 * @returns {Signature} Transaction signature.
	 */
	signTransaction(transaction) {
		return this._facade.signTransaction(this.keyPair, transaction);
	}

	/**
	 * Cosigns a Symbol transaction.
	 * @param {sc.Transaction} transaction Transaction object.
	 * @param {boolean} detached \c true if resulting cosignature is appropriate for network propagation.
	 *                           \c false if resulting cosignature is appropriate for attaching to an aggregate.
	 * @returns {sc.Cosignature|sc.DetachedCosignature} Signed cosignature.
	 */
	cosignTransaction(transaction, detached = false) {
		return this._facade.cosignTransaction(this.keyPair, transaction, detached);
	}

	/**
	 * Cosigns a Symbol transaction hash.
	 * @param {Hash256} transactionHash Transaction hash.
	 * @param {boolean} detached \c true if resulting cosignature is appropriate for network propagation.
	 *                           \c false if resulting cosignature is appropriate for attaching to an aggregate.
	 * @returns {sc.Cosignature|sc.DetachedCosignature} Signed cosignature.
	 */
	cosignTransactionHash(transactionHash, detached = false) {
		return this._facade.static.cosignTransactionHash(this.keyPair, transactionHash, detached);
	}
}

// endregion

/**
 * Facade used to interact with Symbol blockchain.
 */
export class SymbolFacade {
	/**
	 * BIP32 curve name.
	 * @type {string}
	 */
	static BIP32_CURVE_NAME = 'ed25519';

	/**
	 * Network address class type.
	 * @type {typeof Address}
	 */
	static Address = Address;

	/**
	 * Network key pair class type.
	 * @type {typeof KeyPair}
	 */
	static KeyPair = KeyPair;

	/**
	 * Network verifier class type.
	 * @type {typeof Verifier}
	 */
	static Verifier = Verifier;

	/**
	 * Derives shared key from key pair and other party's public key.
	 * @param {KeyPair} keyPair Key pair.
	 * @param {PublicKey} otherPublicKey Other party's public key.
	 * @returns {SharedKey256} Shared encryption key.
	 */
	static deriveSharedKey = deriveSharedKey;

	/**
	 * Creates a Symbol facade.
	 * @param {string|Network} network Symbol network or network name.
	 */
	constructor(network) {
		/**
		 * Underlying network.
		 * @type {Network}
		 */
		this.network = 'string' === typeof network ? NetworkLocator.findByName(Network.NETWORKS, network) : network;

		/**
		 * Underlying transaction factory.
		 * @type {TransactionFactory}
		 */
		this.transactionFactory = new TransactionFactory(this.network);
	}

	/**
	 * Gets class type.
	 * @returns {typeof SymbolFacade} Class type.
	 */
	get static() { // eslint-disable-line class-methods-use-this
		return SymbolFacade;
	}

	/**
	 * Creates a network timestamp representing the current time.
	 * @returns {NetworkTimestamp} Network timestamp representing the current time.
	 */
	now() {
		return this.network.fromDatetime(new Date());
	}

	/**
	 * Creates a Symbol public account from a public key.
	 * @param {PublicKey} publicKey Account public key.
	 * @returns {SymbolPublicAccount} Symbol public account.
	 */
	createPublicAccount(publicKey) {
		return new SymbolPublicAccount(this, publicKey);
	}

	/**
	 * Creates a Symbol account from a private key.
	 * @param {PrivateKey} privateKey Account private key.
	 * @returns {SymbolAccount} Symbol account.
	 */
	createAccount(privateKey) {
		return new SymbolAccount(this, new KeyPair(privateKey));
	}

	/**
	 * Creates a transaction from a (typed) transaction descriptor.
	 * @param {object} typedDescriptor Transaction (typed) descriptor.
	 * @param {PublicKey} signerPublicKey Signer public key.
	 * @param {number} feeMultiplier Fee multiplier.
	 * @param {number} deadlineSeconds Approximate seconds from now for deadline.
	 * @param {number} cosignatureCount Number of cosignature spaces to reserve.
	 * @returns {sc.Transaction} Created transaction.
	 */
	createTransactionFromTypedDescriptor(typedDescriptor, signerPublicKey, feeMultiplier, deadlineSeconds, cosignatureCount = 0) {
		const rawDescriptor = typedDescriptor.toMap();
		const transaction = this.transactionFactory.create({
			...rawDescriptor,

			signerPublicKey,
			deadline: this.now().addSeconds(deadlineSeconds).timestamp
		});

		// if cosignatures are specified in the descriptor, use the max of them and cosignatureCount
		let cosignatureCountAdjustment = cosignatureCount;
		if (rawDescriptor.cosignatures) {
			cosignatureCountAdjustment = rawDescriptor.cosignatures.length > cosignatureCount
				? 0
				: cosignatureCount - rawDescriptor.cosignatures.length;
		}

		const transactionWithCosignaturesSize = transaction.size + (cosignatureCountAdjustment * new sc.Cosignature().size);
		transaction.fee = new sc.Amount(BigInt(transactionWithCosignaturesSize) * BigInt(feeMultiplier));
		return transaction;
	}

	/**
	 * Creates an embedded transaction from a (typed) transaction descriptor.
	 * @param {object} typedDescriptor Transaction (typed) descriptor.
	 * @param {PublicKey} signerPublicKey Signer public key.
	 * @returns {sc.EmbeddedTransaction} Created embedded transaction.
	 */
	createEmbeddedTransactionFromTypedDescriptor(typedDescriptor, signerPublicKey) {
		const transaction = this.transactionFactory.createEmbedded({
			...typedDescriptor.toMap(),

			signerPublicKey
		});
		return transaction;
	}

	/**
	 * Hashes a Symbol transaction.
	 * @param {sc.Transaction} transaction Transaction object.
	 * @returns {Hash256} Transaction hash.
	 */
	hashTransaction(transaction) {
		const hasher = sha3_256.create();
		hasher.update(transaction.signature.bytes);
		hasher.update(transaction.signerPublicKey.bytes);
		hasher.update(this.network.generationHashSeed.bytes);
		hasher.update(transactionDataBuffer(transaction.serialize()));
		return new Hash256(hasher.digest());
	}

	/**
	 * Gets the payload to sign given a Symbol transaction.
	 * @param {sc.Transaction} transaction Transaction object.
	 * @returns {Uint8Array} Verifiable data to sign.
	 */
	extractSigningPayload(transaction) {
		return new Uint8Array([
			...this.network.generationHashSeed.bytes,
			...transactionDataBuffer(transaction.serialize())
		]);
	}

	/**
	 * Signs a Symbol transaction.
	 * @param {KeyPair} keyPair Key pair.
	 * @param {sc.Transaction} transaction Transaction object.
	 * @returns {Signature} Transaction signature.
	 */
	signTransaction(keyPair, transaction) {
		return keyPair.sign(this.extractSigningPayload(transaction));
	}

	/**
	 * Verifies a Symbol transaction.
	 * @param {sc.Transaction} transaction Transaction object.
	 * @param {Signature} signature Signature to verify.
	 * @returns {boolean} \c true if transaction signature is verified.
	 */
	verifyTransaction(transaction, signature) {
		const verifyBuffer = new Uint8Array(this.extractSigningPayload(transaction));
		return new Verifier(transaction.signerPublicKey).verify(verifyBuffer, signature);
	}

	/**
	 * Cosigns a Symbol transaction hash.
	 * @param {KeyPair} keyPair Key pair of the cosignatory.
	 * @param {Hash256} transactionHash Transaction hash.
	 * @param {boolean} detached \c true if resulting cosignature is appropriate for network propagation.
	 *                           \c false if resulting cosignature is appropriate for attaching to an aggregate.
	 * @returns {sc.Cosignature|sc.DetachedCosignature} Signed cosignature.
	 */
	static cosignTransactionHash(keyPair, transactionHash, detached = false) {
		const initializeCosignature = cosignature => {
			cosignature.version = 0n;
			cosignature.signerPublicKey = new sc.PublicKey(keyPair.publicKey.bytes);
			cosignature.signature = new sc.Signature(keyPair.sign(transactionHash.bytes).bytes);
		};

		if (detached) {
			const cosignature = new sc.DetachedCosignature();
			cosignature.parentHash = new sc.Hash256(transactionHash.bytes);
			initializeCosignature(cosignature);
			return cosignature;
		}

		const cosignature = new sc.Cosignature();
		initializeCosignature(cosignature);
		return cosignature;
	}

	/**
	 * Cosigns a Symbol transaction.
	 * @param {KeyPair} keyPair Key pair of the cosignatory.
	 * @param {sc.Transaction} transaction Transaction object.
	 * @param {boolean} detached \c true if resulting cosignature is appropriate for network propagation.
	 *                           \c false if resulting cosignature is appropriate for attaching to an aggregate.
	 * @returns {sc.Cosignature|sc.DetachedCosignature} Signed cosignature.
	 */
	cosignTransaction(keyPair, transaction, detached = false) {
		const transactionHash = this.hashTransaction(transaction);

		return SymbolFacade.cosignTransactionHash(keyPair, transactionHash, detached);
	}

	/**
	 * Hashes embedded transactions of an aggregate transaction.
	 * @param {Array<sc.EmbeddedTransaction>} embeddedTransactions Embedded transactions to hash.
	 * @returns {Hash256} Aggregate transactions hash.
	 */
	static hashEmbeddedTransactions(embeddedTransactions) {
		const hashBuilder = new MerkleHashBuilder();
		embeddedTransactions.forEach(embeddedTransaction => {
			hashBuilder.update(new Hash256(sha3_256(embeddedTransaction.serialize())));
		});

		return hashBuilder.final();
	}

	/**
	 * Creates a network compatible BIP32 path for the specified account.
	 * @param {number} accountId Id of the account for which to generate a BIP32 path.
	 * @returns {Array<number>} BIP32 path for the specified account.
	 */
	bip32Path(accountId) {
		return [44, 'mainnet' === this.network.name ? 4343 : 1, accountId, 0, 0];
	}

	/**
	 * Derives a Symbol KeyPair from a BIP32 node.
	 * @param {Bip32Node} bip32Node BIP32 node.
	 * @returns {KeyPair} Derived key pair.
	 */
	static bip32NodeToKeyPair(bip32Node) {
		return new KeyPair(new PrivateKey(bip32Node.privateKey.bytes));
	}
}
